<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Get Data</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for seqGetData {SeqArray}"><tr><td>seqGetData {SeqArray}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Get Data</h2>

<h3>Description</h3>

<p>Gets data from a sequence GDS file.
</p>


<h3>Usage</h3>

<pre>
seqGetData(gdsfile, var.name)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>gdsfile</code></td>
<td>
<p>a <code><a href="SeqVarGDSClass-class.html">SeqVarGDSClass</a></code> object</p>
</td></tr>
<tr valign="top"><td><code>var.name</code></td>
<td>
<p>the variable name, see details</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The variable name should be <code>"sample.id"</code>, <code>"variant.id"</code>,
<code>"position"</code>, <code>"chromosome"</code>, <code>"allele"</code>, <code>"genotype"</code>,
<code>"annotation/id"</code>, <code>"annotation/qual"</code>, <code>"annotation/filter"</code>,
<code>"annotation/info/VARIABLE_NAME"</code>, or
<code>"annotation/format/VARIABLE_NAME"</code>.
</p>
<p><code>"@genotype"</code>, <code>"annotation/info/@VARIABLE_NAME"</code> or
<code>"annotation/format/@VARIABLE_NAME"</code> are used to obtain the index
associated with these variables.
</p>


<h3>Value</h3>

<p>Return vectors or lists.
</p>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>See Also</h3>

<p><code><a href="seqSetFilter.html">seqSetFilter</a></code>, <code><a href="seqApply.html">seqApply</a></code>
</p>


<h3>Examples</h3>

<pre>
# the GDS file
(gds.fn &lt;- seqExampleFileName("gds"))

# display
(f &lt;- seqOpen(gds.fn))

# get 'sample.id
(samp.id &lt;- seqGetData(f, "sample.id"))
# "NA06984" "NA06985" "NA06986" ...

# get 'variant.id'
head(variant.id &lt;- seqGetData(f, "variant.id"))

# get 'chromosome'
table(seqGetData(f, "chromosome"))

# get 'allele'
head(seqGetData(f, "allele"))
# "T,C" "G,A" "G,A" ...


# set sample and variant filters
seqSetFilter(f, sample.id=samp.id[c(2,4,6,8,10)])
set.seed(100)
seqSetFilter(f, variant.id=sample(variant.id, 10))

# get genotypic data
seqGetData(f, "genotype")

# get annotation/info/DP
seqGetData(f, "annotation/info/DP")

# get annotation/info/AA, a variable-length dataset
seqGetData(f, "annotation/info/AA")
# $length              &lt;- indicating the length of each variable-length data
# [1] 1 1 1 1 1 1 ...
# $data                &lt;- the data according to $length
# [1] "T" "C" "T" "C" "G" "C" ...

# get annotation/format/DP, a variable-length dataset
seqGetData(f, "annotation/format/DP")
# $length              &lt;- indicating the length of each variable-length data
# [1] 1 1 1 1 1 1 ...
# $data                &lt;- the data according to $length
#      variant
# sample [,1] [,2] [,3] [,4] [,5] [,6] ...
#  [1,]   25   25   22    3    4   17  ...

# close the GDS file
seqClose(f)
</pre>

<hr /><div style="text-align: center;">[Package <em>SeqArray</em> version 1.10.2 <a href="00Index.html">Index</a>]</div>
</body></html>
