<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Apply functions via a sliding window over variants</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for seqSlidingWindow {SeqArray}"><tr><td>seqSlidingWindow {SeqArray}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Apply functions via a sliding window over variants</h2>

<h3>Description</h3>

<p>Returns a vector or list of values obtained by applying a function to a
sliding window over variants
</p>


<h3>Usage</h3>

<pre>
seqSlidingWindow(gdsfile, var.name, win.size, shift=1, FUN,
    as.is = c("list", "integer", "double", "character", "none"),
    var.index = c("none", "relative", "absolute"), ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>gdsfile</code></td>
<td>
<p>a <code><a href="SeqVarGDSClass-class.html">SeqVarGDSClass</a></code> object</p>
</td></tr>
<tr valign="top"><td><code>var.name</code></td>
<td>
<p>the variable name(s), see details</p>
</td></tr>
<tr valign="top"><td><code>win.size</code></td>
<td>
<p>the size of sliding window</p>
</td></tr>
<tr valign="top"><td><code>shift</code></td>
<td>
<p>the number of variants to shift the window at each step</p>
</td></tr>
<tr valign="top"><td><code>FUN</code></td>
<td>
<p>the function to be applied</p>
</td></tr>
<tr valign="top"><td><code>as.is</code></td>
<td>
<p>returned value: a list, an integer vector, etc</p>
</td></tr>
<tr valign="top"><td><code>var.index</code></td>
<td>
<p>if <code>"none"</code>, call <code>FUN(x, ...)</code> without variable
index; if <code>"relative"</code> or <code>"absolute"</code>, add an argument to
the user-defined function <code>FUN</code> like <code>FUN(index, x, ...)</code>
where <code>index</code> is an index of variant starting from 1:
<code>"relative"</code> for indexing in the selection defined by
<code><a href="seqSetFilter.html">seqSetFilter</a></code>, <code>"absolute"</code> for indexing with
respect to all data</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>optional arguments to <code>FUN</code></p>
</td></tr>
</table>


<h3>Details</h3>

<p>The variable name should be <code>"sample.id"</code>, <code>"variant.id"</code>,
<code>"position"</code>, <code>"chromosome"</code>, <code>"allele"</code>, <code>"annotation/id"</code>,
<code>"annotation/qual"</code>, <code>"annotation/filter"</code>,
<code>"annotation/info/VARIABLE_NAME"</code>, or <code>"annotation/format/VARIABLE_NAME"</code>.
</p>
<p>In the user-defined funciton <code>FUN(x, ...)</code> or
<code>FUN(index, x, ...)</code>, <code>x</code> is a list with <code>win.size</code> elements,
and each element includes values for the variable(s) <code>var.name</code>;
<code>index</code> is the starting position of the sliding window.
</p>
<p>The algorithm is highly optimized by blocking the computations to exploit the
high-speed memory instead of disk.
</p>


<h3>Value</h3>

<p>A vector or list of values.
</p>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>See Also</h3>

<p><code><a href="seqSetFilter.html">seqSetFilter</a></code>, <code><a href="seqGetData.html">seqGetData</a></code>,
<code><a href="seqApply.html">seqApply</a></code>
</p>


<h3>Examples</h3>

<pre>
# the file of GDS
gds.fn &lt;- seqExampleFileName("gds")
# or gds.fn &lt;- "C:/YourFolder/Your_GDS_File.gds"

# display
(f &lt;- seqOpen(gds.fn))

# get 'sample.id
(samp.id &lt;- seqGetData(f, "sample.id"))
# "NA06984" "NA06985" "NA06986" ...

# get 'variant.id'
head(variant.id &lt;- seqGetData(f, "variant.id"))


# set sample and variant filters
set.seed(100)
seqSetFilter(f, sample.id=samp.id[seq(2, 16, 2)],
	variant.id=sample(variant.id, 10))


# apply a function via a sliding window over variants
seqSlidingWindow(f, c(qual="annotation/id"), win.size=3,
	FUN = function(x) {
		# x is a list with 'win.size' elements
		print(x)
	}, as.is="none")

# apply a function via a sliding window over variants
seqSlidingWindow(f, c(qual="annotation/id"), win.size=3,
	FUN = function(x) {
		cat(unlist(x), sep="\t"); cat("\n")
	}, as.is="none")


# apply a function via a sliding window over variants
seqSlidingWindow(f, c(geno="genotype", phase="phase", qual="annotation/id"),
	FUN = function(index, x) {
		cat("Window ", index, ":\n", sep="")
		print(x)
	},
	win.size=3, as.is="none", var.index="relative")


# apply a function via a sliding window over variants
seqSlidingWindow(f, "genotype", win.size=4,
	FUN = function(index, x) {
		z &lt;- unlist(lapply(x, function(z) mean(z, na.rm=TRUE)))
		cat("Window ", index, ", starting from Variant ", index,
			"\n    ", format(round(z,3), nsmall=3, width=8), "\n", sep="")
	},
	as.is="none", var.index="relative")


# close the GDS file
seqClose(f)
</pre>

<hr /><div style="text-align: center;">[Package <em>SeqArray</em> version 1.8.0 <a href="00Index.html">Index</a>]</div>
</body></html>
