<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Reformat VCF Files</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for seqVCF2GDS {SeqArray}"><tr><td>seqVCF2GDS {SeqArray}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Reformat VCF Files</h2>

<h3>Description</h3>

<p>Reformats Variant Call Format (VCF) files.
</p>


<h3>Usage</h3>

<pre>
seqVCF2GDS(vcf.fn, out.fn, header=NULL, storage.option="ZIP_RA",
    info.import=NULL, fmt.import=NULL, genotype.var.name="GT",
    ignore.chr.prefix="chr", reference=NULL, start=1L, count=-1L,
    optimize=TRUE, raise.error=TRUE, digest=TRUE, parallel=FALSE,
    verbose=TRUE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>vcf.fn</code></td>
<td>
<p>the file name(s) of VCF format; or a <code><a href="../../base/html/connections.html">connection</a></code>
object</p>
</td></tr>
<tr valign="top"><td><code>out.fn</code></td>
<td>
<p>the file name of output GDS file</p>
</td></tr>
<tr valign="top"><td><code>header</code></td>
<td>
<p>if NULL, <code>header</code> is set to be
<code><a href="seqVCF_Header.html">seqVCF_Header</a>(vcf.fn)</code></p>
</td></tr>
<tr valign="top"><td><code>storage.option</code></td>
<td>
<p>specify the storage and compression option,
&quot;ZIP_RA&quot; (<code><a href="seqStorageOption.html">seqStorageOption</a>("ZIP_RA")</code>) by default; or
&quot;LZMA_RA&quot; to use LZMA compression algorithm with higher compression
ratio</p>
</td></tr>
<tr valign="top"><td><code>info.import</code></td>
<td>
<p>characters, the variable name(s) in the INFO field
for import; or <code>NULL</code> for all variables</p>
</td></tr>
<tr valign="top"><td><code>fmt.import</code></td>
<td>
<p>characters, the variable name(s) in the FORMAT field
for import; or <code>NULL</code> for all variables</p>
</td></tr>
<tr valign="top"><td><code>genotype.var.name</code></td>
<td>
<p>the ID for genotypic data in the FORMAT column;
&quot;GT&quot; by default, VCFv4.0</p>
</td></tr>
<tr valign="top"><td><code>ignore.chr.prefix</code></td>
<td>
<p>a vector of character, indicating the prefix of
chromosome which should be ignored, like &quot;chr&quot;; it is not
case-sensitive</p>
</td></tr>
<tr valign="top"><td><code>reference</code></td>
<td>
<p>genome reference, like &quot;hg19&quot;, &quot;GRCh37&quot;; if the genome
reference is not available in VCF files, users could specify the
reference here</p>
</td></tr>
<tr valign="top"><td><code>start</code></td>
<td>
<p>the starting variant if importing part of VCF files</p>
</td></tr>
<tr valign="top"><td><code>count</code></td>
<td>
<p>the maximum count of variant if importing part of VCF files,
-1 indicates importing to the end</p>
</td></tr>
<tr valign="top"><td><code>optimize</code></td>
<td>
<p>if <code>TRUE</code>, optimize the access efficiency by calling
<code><a href="../../gdsfmt/html/cleanup.gds.html">cleanup.gds</a></code></p>
</td></tr>
<tr valign="top"><td><code>raise.error</code></td>
<td>
<p><code>TRUE</code>: throw an error if numeric conversion fails;
<code>FALSE</code>: get missing value if numeric conversion fails</p>
</td></tr>
<tr valign="top"><td><code>digest</code></td>
<td>
<p>a logical value (TRUE/FALSE) or a character (&quot;md5&quot;, &quot;sha1&quot;,
&quot;sha256&quot;, &quot;sha384&quot; or &quot;sha512&quot;); add md5 hash codes to the GDS file
if TRUE or a digest algorithm is specified</p>
</td></tr>
<tr valign="top"><td><code>parallel</code></td>
<td>
<p><code>FALSE</code> (serial processing), <code>TRUE</code> (parallel
processing), a numeric value indicating the number of cores, or a
cluster object for parallel processing; <code>parallel</code> is passed to
the argument <code>cl</code> in <code><a href="seqParallel.html">seqParallel</a></code>, see
<code><a href="seqParallel.html">seqParallel</a></code> for more details</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>if <code>TRUE</code>, show information</p>
</td></tr>
</table>


<h3>Details</h3>

<p>If there are more than one files in <code>vcf.fn</code>, <code>seqVCF2GDS</code> will
merge all VCF files together if they contain the same samples. It is useful
to merge multiple VCF files if variant data are split by chromosomes.
</p>
<p>The real numbers in the VCF file(s) are stored in 32-bit floating-point
format by default. Users can set
<code>storage.option=seqStorageOption(float.mode="float64")</code>
to switch to 64-bit floating point format. Or packed real numbers can be
adopted by setting
<code>storage.option=seqStorageOption(float.mode="packedreal16:scale=0.0001")</code>.
</p>
<p>By default, the compression method is &quot;ZIP_RA&quot; (zlib algorithm with default
compression level + independent data blocks). Users can maximize the
compression ratio by <code>storage.option="ZIP_RA.max"</code> or
<code>storage.option=seqStorageOption("ZIP_RA.max")</code>.
LZ4 (<a href="http://cyan4973.github.io/lz4/">http://cyan4973.github.io/lz4/</a>) is an option via
<code>storage.option="LZ4_RA"</code> or
<code>storage.option=seqStorageOption("LZ4_RA")</code>.
LZMA (xz, <a href="http://tukaani.org/xz/">http://tukaani.org/xz/</a>) is another option via
<code>storage.option="LZMA_RA"</code> or
<code>storage.option=seqStorageOption("LZMA_RA")</code>, and it is known to have
higher compression ratio than zlib.
</p>
<p>If multiple cores/processes are specified in <code>parallel</code>, all VCF files
are scanned to calculate the total number of variants before format conversion.
</p>


<h3>Value</h3>

<p>Return the file name of GDS format with an absolute path.
</p>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>References</h3>

<p>Danecek, P., Auton, A., Abecasis, G., Albers, C.A., Banks, E., DePristo,
M.A., Handsaker, R.E., Lunter, G., Marth, G.T., Sherry, S.T., et al. (2011).
The variant call format and VCFtools. Bioinformatics 27, 2156-2158.
</p>


<h3>See Also</h3>

<p><code><a href="seqVCF_Header.html">seqVCF_Header</a></code>, <code><a href="seqStorageOption.html">seqStorageOption</a></code>,
<code><a href="seqMerge.html">seqMerge</a></code>, <code><a href="seqGDS2VCF.html">seqGDS2VCF</a></code>
</p>


<h3>Examples</h3>

<pre>
# the VCF file
vcf.fn &lt;- seqExampleFileName("vcf")

# conversion
seqVCF2GDS(vcf.fn, "tmp.gds")

# conversion in parallel
seqVCF2GDS(vcf.fn, "tmp_p2.gds", parallel=2L)


# display
(f &lt;- seqOpen("tmp.gds"))
seqClose(f)



# convert without the INFO fields
seqVCF2GDS(vcf.fn, "tmp.gds", info.import=character(0))

# display
(f &lt;- seqOpen("tmp.gds"))
seqClose(f)



# convert without the INFO and FORMAT fields
seqVCF2GDS(vcf.fn, "tmp.gds", info.import=character(0), fmt.import=character(0))

# display
(f &lt;- seqOpen("tmp.gds"))
seqClose(f)


# delete the temporary file
unlink(c("tmp.gds", "tmp_p2.gds"), force=TRUE)
</pre>

<hr /><div style="text-align: center;">[Package <em>SeqArray</em> version 1.12.9 <a href="00Index.html">Index</a>]</div>
</body></html>
