<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Apply Functions Over Array Margins</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for seqApply {SeqArray}"><tr><td>seqApply {SeqArray}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Apply Functions Over Array Margins</h2>

<h3>Description</h3>

<p>Returns a vector or list of values obtained by applying a function to
margins of genotypes and annotations.
</p>


<h3>Usage</h3>

<pre>
seqApply(gdsfile, var.name, FUN, margin=c("by.variant", "by.sample"),
    as.is=c("none", "list", "integer", "double", "character", "logical", "raw"),
    var.index=c("none", "relative", "absolute"), parallel=FALSE,
    .useraw=FALSE, .progress=FALSE, .list_dup=TRUE, ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>gdsfile</code></td>
<td>
<p>a <code><a href="SeqVarGDSClass-class.html">SeqVarGDSClass</a></code> object</p>
</td></tr>
<tr valign="top"><td><code>var.name</code></td>
<td>
<p>the variable name(s), see details</p>
</td></tr>
<tr valign="top"><td><code>FUN</code></td>
<td>
<p>the function to be applied</p>
</td></tr>
<tr valign="top"><td><code>margin</code></td>
<td>
<p>giving the dimension which the function will be applied over</p>
</td></tr>
<tr valign="top"><td><code>as.is</code></td>
<td>
<p>returned value: a list, an integer vector, etc;
<code>as.is</code> can be a <code><a href="../../base/html/connections.html">connection</a></code> object, or a GDS node
<code><a href="../../gdsfmt/html/gdsn.class.html">gdsn.class</a></code> object; if &quot;unlist&quot; is used, produces a vector
which contains all the atomic components, via
<code>unlist(..., recursive=FALSE)</code></p>
</td></tr>
<tr valign="top"><td><code>var.index</code></td>
<td>
<p>if <code>"none"</code>, call <code>FUN(x, ...)</code> without variable
index; if <code>"relative"</code> or <code>"absolute"</code>, add an argument to
the user-defined function <code>FUN</code> like <code>FUN(index, x, ...)</code>
where <code>index</code> is an index of variant starting from 1 if
<code>margin = "by.variant"</code>: <code>"relative"</code> for indexing in the
selection defined by <code><a href="seqSetFilter.html">seqSetFilter</a></code>, <code>"absolute"</code>
for indexing with respect to all data</p>
</td></tr>
<tr valign="top"><td><code>parallel</code></td>
<td>
<p><code>FALSE</code> (serial processing), <code>TRUE</code> (multicore
processing), numeric value or other value; <code>parallel</code> is passed
to the argument <code>cl</code> in <code><a href="seqParallel.html">seqParallel</a></code>, see
<code><a href="seqParallel.html">seqParallel</a></code> for more details.</p>
</td></tr>
<tr valign="top"><td><code>.useraw</code></td>
<td>
<p><code>TRUE</code>, force to use RAW instead of INTEGER for
genotypes and dosages; <code>FALSE</code>, use INTEGER; <code>NA</code>, use RAW
instead of INTEGER if possible</p>
</td></tr>
<tr valign="top"><td><code>.progress</code></td>
<td>
<p>if <code>TRUE</code>, show progress information</p>
</td></tr>
<tr valign="top"><td><code>.list_dup</code></td>
<td>
<p>internal use only</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>optional arguments to <code>FUN</code></p>
</td></tr>
</table>


<h3>Details</h3>

<p>The variable name should be <code>"sample.id"</code>, <code>"variant.id"</code>,
<code>"position"</code>, <code>"chromosome"</code>, <code>"allele"</code>, <code>"genotype"</code>,
<code>"annotation/id"</code>, <code>"annotation/qual"</code>, <code>"annotation/filter"</code>,
<code>"annotation/info/VARIABLE_NAME"</code>, or
<code>"annotation/format/VARIABLE_NAME"</code>.
</p>
<p><code>"$dosage"</code> is also allowed for the dosages of reference allele (integer:
0, 1, 2 and NA for diploid genotypes).
</p>
<p><code>"$num_allele"</code> returns an integer vector with the numbers of distinct
alleles.
</p>
<p>The algorithm is highly optimized by blocking the computations to exploit
the high-speed memory instead of disk.
</p>


<h3>Value</h3>

<p>A vector, a list of values or none.
</p>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>See Also</h3>

<p><code><a href="seqBlockApply.html">seqBlockApply</a></code>, <code><a href="seqSetFilter.html">seqSetFilter</a></code>,
<code><a href="seqGetData.html">seqGetData</a></code>, <code><a href="seqParallel.html">seqParallel</a></code>,
<code><a href="seqParallelSetup.html">seqGetParallel</a></code>
</p>


<h3>Examples</h3>

<pre>
# the GDS file
(gds.fn &lt;- seqExampleFileName("gds"))

# display
(f &lt;- seqOpen(gds.fn))

# get 'sample.id
(samp.id &lt;- seqGetData(f, "sample.id"))
# "NA06984" "NA06985" "NA06986" ...

# get 'variant.id'
head(variant.id &lt;- seqGetData(f, "variant.id"))


# set sample and variant filters
set.seed(100)
seqSetFilter(f, sample.id=samp.id[c(2,4,6,8,10)],
    variant.id=sample(variant.id, 10))

# read
seqApply(f, "genotype", FUN=print, margin="by.variant")
seqApply(f, "genotype", FUN=print, margin="by.variant", .useraw=TRUE)

seqApply(f, "genotype", FUN=print, margin="by.sample")
seqApply(f, "genotype", FUN=print, margin="by.sample", .useraw=TRUE)


# read multiple variables variant by variant
seqApply(f, c(geno="genotype", phase="phase", qual="annotation/id",
    DP="annotation/format/DP"), FUN=print, as.is="none")

# get the numbers of alleles per variant
seqApply(f, "allele",
    FUN=function(x) length(unlist(strsplit(x,","))), as.is="integer")

# output to a file
fl &lt;- file("tmp.txt", "wt")
seqApply(f, "genotype", FUN=sum, na.rm=TRUE, as.is=fl)
close(fl)
readLines("tmp.txt")

seqApply(f, "genotype", FUN=sum, na.rm=TRUE, as.is=stdout())
seqApply(f, "genotype", FUN=sum, na.rm=TRUE, as.is="integer")
# should be identical



################################################################
# with an index of variant

seqApply(f, c(geno="genotype", phase="phase", qual="annotation/id"),
    FUN=function(index, x) { print(index); print(x); index },
    as.is="integer", var.index="relative")
# it is as the same as
which(seqGetFilter(f)$variant.sel)



################################################################
# reset sample and variant filters
seqResetFilter(f)

# calculate the frequency of reference allele,
#   a faster version could be obtained by C coding
af &lt;- seqApply(f, "genotype", FUN=function(x) mean(x==0L, na.rm=TRUE),
    as.is="double")
length(af)
summary(af)



################################################################
# apply the user-defined function sample by sample

# reset sample and variant filters
seqResetFilter(f)
summary(seqApply(f, "genotype", FUN=function(x) { mean(is.na(x)) },
    margin="by.sample", as.is="double"))

# set sample and variant filters
set.seed(100)
seqSetFilter(f, sample.id=samp.id[c(2,4,6,8,10)],
    variant.id=sample(variant.id, 10))

seqApply(f, "genotype", FUN=print, margin="by.variant", as.is="none")

seqApply(f, "genotype", FUN=print, margin="by.sample", as.is="none")

seqApply(f, c(sample.id="sample.id", genotype="genotype"), FUN=print,
    margin="by.sample", as.is="none")


# close the GDS file
seqClose(f)


# delete the temporary file
unlink("tmp.txt")
</pre>

<hr /><div style="text-align: center;">[Package <em>SeqArray</em> version 1.13.6 <a href="00Index.html">Index</a>]</div>
</body></html>
