<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Add a Variable to the INFO Field</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for seqInfoNewVar {SeqArray}"><tr><td>seqInfoNewVar {SeqArray}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Add a Variable to the INFO Field</h2>

<h3>Description</h3>

<p>Adds a new variable to the INFO field in the specified GDS file.
</p>


<h3>Usage</h3>

<pre>
seqInfoNewVar(gdsfile, var.name, variant.id, val,
    description="", compress=c("ZIP.MAX", ""), no.data.index=TRUE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>gdsfile</code></td>
<td>
<p>a <code><a href="SeqVarGDSClass-class.html">SeqVarGDSClass</a></code> object</p>
</td></tr>
<tr valign="top"><td><code>var.name</code></td>
<td>
<p>the variable name(s), see details</p>
</td></tr>
<tr valign="top"><td><code>variant.id</code></td>
<td>
<p>the variant ID(s), should have the same order as
IDs in the GDS file</p>
</td></tr>
<tr valign="top"><td><code>val</code></td>
<td>
<p>a vector or a matrix with the same order as <code>variant.id</code>;
if it is a matrix, the number of columns should be equal to the
length of <code>variant.id</code>, see details</p>
</td></tr>
<tr valign="top"><td><code>description</code></td>
<td>
<p>the variable description</p>
</td></tr>
<tr valign="top"><td><code>compress</code></td>
<td>
<p>to specify the compression algorithm: <code>""</code>
for no compression; see <code>compress</code> in the function
<code><a href="../../gdsfmt/html/add.gdsn.html">add.gdsn</a></code></p>
</td></tr>
<tr valign="top"><td><code>no.data.index</code></td>
<td>
<p>applicable only if <code>val</code> is a numeric vector
or a factor variable; if <code>no.data.index=TRUE</code>, non-exist values
in <code>val</code> will be replaced by <code>NA</code> or <code>NaN</code> with respect
to the variants not specified in <code>variant.id</code>, and no index
data associated with this variable are created</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The variable name should be <code>"sample.id"</code>, <code>"variant.id"</code>,
<code>"position"</code>, <code>"chromosome"</code>, <code>"allele"</code>, <code>"annotation/id"</code>,
<code>"annotation/qual"</code>, <code>"annotation/filter"</code>,
<code>"annotation/info/VARIABLE_NAME"</code>, or <code>"annotation/format/VARIABLE_NAME"</code>.
</p>
<p>The argument <code>val</code> should be integers, numeric values, a logical
variable, characters or factors. If <code>val</code> is a logical variable, one-bit
storage mode will be used to store this variable, which corresponds to the
variable defined with 'Type=Flag' in the VCF format.
</p>


<h3>Value</h3>

<p>None.
</p>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>See Also</h3>

<p><code><a href="seqGetData.html">seqGetData</a></code>
</p>


<h3>Examples</h3>

<pre>
# the GDS file
gds.fn &lt;- seqExampleFileName("gds")

file.copy(gds.fn, "test.gds", overwrite=TRUE)

# display
(f &lt;- seqOpen("test.gds", readonly=FALSE))

# get variant IDs
variant.id &lt;- seqGetData(f, "variant.id")


####    add variables to the INFO field    ####

set.seed(100)

seqInfoNewVar(f, "int", variant.id[1:15], sample.int(256, 15, TRUE),
    "integer variable")
seqGetData(f, "annotation/info/int")

seqInfoNewVar(f, "int.2", variant.id[1:15], sample.int(256, 15, TRUE),
    "integer variable", no.data.index=FALSE)
seqGetData(f, "annotation/info/int.2")


seqInfoNewVar(f, "numeric", variant.id[15:30], rnorm(16),
    "numeric variable")
seqGetData(f, "annotation/info/numeric")

seqInfoNewVar(f, "numeric.2", variant.id[15:30], rnorm(16),
    "numeric variable", no.data.index=FALSE)
seqGetData(f, "annotation/info/numeric.2")


seqInfoNewVar(f, "flag", variant.id[4:9], rep(c(FALSE, TRUE), 3),
    "flag variable")
# stored in `bit1'
seqGetData(f, "annotation/info/flag")


seqInfoNewVar(f, "factor", variant.id,
    factor(c("ABC", "DDD", "CVX")[sample(1:3, length(variant.id), TRUE)]),
    "string/factor variable")
# stored in `int32' with attributes
seqGetData(f, "annotation/info/factor")


seqInfoNewVar(f, "string", variant.id,
    c("ABC", "DDD", "CVX")[sample(1:3, length(variant.id), TRUE)],
    "string variable")
seqGetData(f, "annotation/info/string")


# show the file
f


# the corresponding VCF file
seqGDS2VCF(f, "test.vcf.gz")
txt &lt;- strsplit(readLines("test.vcf.gz", n=40), "\t")[-c(1:21)]

# the INFO field:
sapply(txt, function(x) x[8])


# close the GDS file
seqClose(f)


# delete the temporary files
unlink("test.gds", force=TRUE)
unlink("test.vcf.gz", force=TRUE)
</pre>

<hr /><div style="text-align: center;">[Package <em>SeqArray</em> version 1.9.13 <a href="00Index.html">Index</a>]</div>
</body></html>
