<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Set a Filter to Variant with Allele Count/Freq</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for seqSetFilterCond {SeqArray}"><tr><td>seqSetFilterCond {SeqArray}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Set a Filter to Variant with Allele Count/Freq</h2>

<h3>Description</h3>

<p>Sets a filter to variant with specified allele count/frequency and
missing rate.
</p>


<h3>Usage</h3>

<pre>
seqSetFilterCond(gdsfile, maf=NaN, mac=1L, missing.rate=NaN,
    parallel=seqGetParallel(), .progress=FALSE, verbose=TRUE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>gdsfile</code></td>
<td>
<p>a <code><a href="SeqVarGDSClass-class.html">SeqVarGDSClass</a></code> object</p>
</td></tr>
<tr valign="top"><td><code>maf</code></td>
<td>
<p>minimum minor reference allele frequency, or a range of MAF
<code>maf[1] &lt;= ... &lt; maf[2]</code></p>
</td></tr>
<tr valign="top"><td><code>mac</code></td>
<td>
<p>minimum minor reference allele count, or a range of MAC
<code>mac[1] &lt;= ... &lt; mac[2]</code></p>
</td></tr>
<tr valign="top"><td><code>missing.rate</code></td>
<td>
<p>maximum missing genotype rate</p>
</td></tr>
<tr valign="top"><td><code>.progress</code></td>
<td>
<p>if <code>TRUE</code>, show progress information</p>
</td></tr>
<tr valign="top"><td><code>parallel</code></td>
<td>
<p><code>FALSE</code> (serial processing), <code>TRUE</code> (multicore
processing), numeric value or other value; <code>parallel</code> is passed
to the argument <code>cl</code> in <code><a href="seqParallel.html">seqParallel</a></code>, see
<code><a href="seqParallel.html">seqParallel</a></code> for more details.</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>if <code>TRUE</code>, show information</p>
</td></tr>
</table>


<h3>Value</h3>

<p>None.
</p>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>See Also</h3>

<p><code><a href="seqSetFilter.html">seqSetFilter</a></code>, <code><a href="seqSetFilter.html">seqSetFilterChrom</a></code>,
<code><a href="seqAlleleFreq.html">seqAlleleFreq</a></code>, <code><a href="seqAlleleFreq.html">seqAlleleCount</a></code>,
<code><a href="seqMissing.html">seqMissing</a></code>
</p>


<h3>Examples</h3>

<pre>
# the GDS file
(gds.fn &lt;- seqExampleFileName("gds"))

# display
(f &lt;- seqOpen(gds.fn))

seqSetFilterChrom(f, c(1, 6))
seqSetFilterCond(f, maf=0.05, .progress=TRUE)

seqSetFilterChrom(f, c(1, 6))
seqSetFilterCond(f, maf=c(0.01, 0.05), .progress=TRUE)



# close the GDS file
seqClose(f)
</pre>

<hr /><div style="text-align: center;">[Package <em>SeqArray</em> version 1.17.1 <a href="00Index.html">Index</a>]</div>
</body></html>
