<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Chromosome Selection</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for seqSetFilterChrom {SeqArray}"><tr><td>seqSetFilterChrom {SeqArray}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Chromosome Selection</h2>

<h3>Description</h3>

<p>Selects the variants according to the specified chromosome(s).
</p>


<h3>Usage</h3>

<pre>
seqSetFilterChrom(gdsfile, include=NULL, is.num=NA, from.bp=NULL, to.bp=NULL,
    verbose=TRUE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>gdsfile</code></td>
<td>
<p>a <code><a href="SeqVarGDSClass-class.html">SeqVarGDSClass</a></code> object</p>
</td></tr>
<tr valign="top"><td><code>include</code></td>
<td>
<p>NULL, or character for specified chromosome(s)</p>
</td></tr>
<tr valign="top"><td><code>is.num</code></td>
<td>
<p>a logical variable: <code>TRUE</code>, chromosome code is numeric;
<code>FALSE</code>, chromosome is not numeric; <code>is.num=TRUE</code> is usually
used to exclude non-autosomes</p>
</td></tr>
<tr valign="top"><td><code>from.bp</code></td>
<td>
<p>NULL, no limit; numeric, the lower bound of position</p>
</td></tr>
<tr valign="top"><td><code>to.bp</code></td>
<td>
<p>NULL, no limit; numeric, the upper bound of position</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>if <code>TRUE</code>, show information</p>
</td></tr>
</table>


<h3>Details</h3>

<p>If <code>from.bp</code> and <code>to.bp</code> has values, they should be equal-size
as <code>include</code>. A trio of <code>include</code>, <code>from.bp</code> and <code>to.bp</code>
indicates a region on human genomes. <code>NA</code> in <code>from.bp</code> is treated
as 0, and <code>NA</code> in <code>to.bp</code> is treated as the maximum of integer
(2^31 - 1).
</p>


<h3>Value</h3>

<p>None.
</p>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>See Also</h3>

<p><code><a href="seqSetFilter.html">seqSetFilter</a></code>
</p>


<h3>Examples</h3>

<pre>
# the GDS file
(gds.fn &lt;- seqExampleFileName("gds"))

f &lt;- seqOpen(gds.fn)
seqSummary(f)


seqSetFilterChrom(f, is.num=TRUE)
seqSummary(f, "genotype", check="none")

seqSetFilterChrom(f, is.num=FALSE)
seqSummary(f, "genotype", check="none")

seqSetFilterChrom(f, 1:4)
seqSummary(f, "genotype", check="none")
table(seqGetData(f, "chromosome"))

# HLA region
seqSetFilterChrom(f, 6, from.bp=29719561, to.bp=32883508)
seqSummary(f, "genotype", check="none")

# two regions
seqSetFilterChrom(f, c(1, 6), from.bp=c(1000000, 29719561),
    to.bp=c(90000000, 32883508))
seqSummary(f, "genotype", check="none")
seqGetData(f, "chromosome")

# close the GDS file
seqClose(f)
</pre>

<hr /><div style="text-align: center;">[Package <em>SeqArray</em> version 1.13.0 <a href="00Index.html">Index</a>]</div>
</body></html>
