<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Set a Filter to Sample or Variant</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for seqSetFilter-methods {SeqArray}"><tr><td>seqSetFilter-methods {SeqArray}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Set a Filter to Sample or Variant</h2>

<h3>Description</h3>

<p>Sets a filter to sample and/or variant.
</p>


<h3>Usage</h3>

<pre>
## S4 method for signature 'SeqVarGDSClass,ANY'
seqSetFilter(object, variant.sel,
    sample.sel=NULL, variant.id=NULL, sample.id=NULL,
    action=c("set", "intersect", "push", "push+set", "push+intersect", "pop"),
    verbose=TRUE)
## S4 method for signature 'SeqVarGDSClass,GRanges'
seqSetFilter(object, variant.sel,
    rm.txt="chr", verbose=TRUE)
## S4 method for signature 'SeqVarGDSClass,GRangesList'
seqSetFilter(object, variant.sel,
    rm.txt="chr", verbose=TRUE)
## S4 method for signature 'SeqVarGDSClass,IRanges'
seqSetFilter(object, variant.sel,
    chr, verbose=TRUE)
seqResetFilter(object, sample=TRUE, variant=TRUE, verbose=TRUE)
seqSetFilterChrom(object, include=NULL, is.num=NA, from.bp=NULL, to.bp=NULL,
    verbose=TRUE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>object</code></td>
<td>
<p>a <code><a href="SeqVarGDSClass-class.html">SeqVarGDSClass</a></code> object</p>
</td></tr>
<tr valign="top"><td><code>variant.sel</code></td>
<td>
<p>a logical/raw/index vector indicating the selected
variants; <code><a href="../../GenomicRanges/html/GRanges-class.html">GRanges</a></code>, a GRanges object for the genomic
locations; <code><a href="../../GenomicRanges/html/GRangesList-class.html">GRangesList</a></code>, a GRangesList object for storing
a collection of GRanges objects; <code><a href="../../IRanges/html/IRanges-constructor.html">IRanges</a></code>, a IRanges
object for storing a collection of range objects</p>
</td></tr>
<tr valign="top"><td><code>sample.sel</code></td>
<td>
<p>a logical/raw/index vector indicating the selected
samples</p>
</td></tr>
<tr valign="top"><td><code>variant.id</code></td>
<td>
<p>ID of selected variants</p>
</td></tr>
<tr valign="top"><td><code>sample.id</code></td>
<td>
<p>ID of selected samples</p>
</td></tr>
<tr valign="top"><td><code>action</code></td>
<td>
<p><code>"set"</code> &ndash; set the current filter via <code>sample.id</code>,
<code>variant.id</code>, <code>samp.sel</code> or <code>variant.sel</code>;
<code>"intersect"</code> &ndash; set the current filter to the intersection of
selected samples and/or variants;
<code>"push"</code> &ndash; push the current filter to the stack, and it could
be recovered by <code>"pop"</code> later, no change on the current
filter;
<code>"push+set"</code> &ndash; push the current filter to the stack, and changes
the current filter via <code>sample.id</code>, <code>variant.id</code>,
<code>samp.sel</code> or <code>variant.sel</code>;
<code>"push+intersect"</code> &ndash; push the current filter to the stack, and
set the current filter to the intersection of selected samples
and/or variants;
<code>"pop"</code> &ndash; pop up the last filter</p>
</td></tr>
<tr valign="top"><td><code>rm.txt</code></td>
<td>
<p>a character, the characters will be removed from
<code>seqnames(variant.sel)</code></p>
</td></tr>
<tr valign="top"><td><code>chr</code></td>
<td>
<p>a vector of character for chromsome coding</p>
</td></tr>
<tr valign="top"><td><code>sample</code></td>
<td>
<p>logical, if <code>TRUE</code>, include all samples</p>
</td></tr>
<tr valign="top"><td><code>variant</code></td>
<td>
<p>logical, if <code>TRUE</code>, include all variants</p>
</td></tr>
<tr valign="top"><td><code>include</code></td>
<td>
<p>NULL, or character for specified chromosome(s)</p>
</td></tr>
<tr valign="top"><td><code>is.num</code></td>
<td>
<p>a logical variable: <code>TRUE</code>, chromosome code is numeric;
<code>FALSE</code>, chromosome is not numeric; <code>is.num=TRUE</code> is usually
used to exclude non-autosomes</p>
</td></tr>
<tr valign="top"><td><code>from.bp</code></td>
<td>
<p>NULL, no limit; numeric, the lower bound of position</p>
</td></tr>
<tr valign="top"><td><code>to.bp</code></td>
<td>
<p>NULL, no limit; numeric, the upper bound of position</p>
</td></tr>
<tr valign="top"><td><code>verbose</code></td>
<td>
<p>if <code>TRUE</code>, show information</p>
</td></tr>
</table>


<h3>Details</h3>

<p><code>seqResetFilter(file)</code> is equivalent to <code>seqSetFilter(file)</code>,
where the selection arguments in <code>seqSetFilter</code> are <code>NULL</code>.
</p>
<p>If <code>from.bp</code> and <code>to.bp</code> has values, they should be equal-size
as <code>include</code>. A trio of <code>include</code>, <code>from.bp</code> and <code>to.bp</code>
indicates a region on human genomes. <code>NA</code> in <code>from.bp</code> is treated
as 0, and <code>NA</code> in <code>to.bp</code> is treated as the maximum of integer
(2^31 - 1).
</p>


<h3>Value</h3>

<p>None.
</p>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>See Also</h3>

<p><code><a href="seqSetFilter.html">seqSetFilterChrom</a></code>, <code><a href="seqGetFilter.html">seqGetFilter</a></code>,
<code><a href="seqGetData.html">seqGetData</a></code>, <code><a href="seqApply.html">seqApply</a></code>
</p>


<h3>Examples</h3>

<pre>
# the GDS file
(gds.fn &lt;- seqExampleFileName("gds"))

# display
(f &lt;- seqOpen(gds.fn))

# get 'sample.id
(samp.id &lt;- seqGetData(f, "sample.id"))
# "NA06984" "NA06985" "NA06986" ...

# get 'variant.id'
head(variant.id &lt;- seqGetData(f, "variant.id"))

# get 'chromosome'
table(seqGetData(f, "chromosome"))

# get 'allele'
head(seqGetData(f, "allele"))
# "T,C" "G,A" "G,A" ...


# set sample and variant filters
seqSetFilter(f, sample.id=samp.id[c(2,4,6,8)])
set.seed(100)
seqSetFilter(f, variant.id=sample(variant.id, 5))

# get genotypic data
seqGetData(f, "genotype")


## OR
# set sample and variant filters
seqSetFilter(f, sample.sel=c(2,4,6,8))
set.seed(100)
seqSetFilter(f, variant.sel=sample.int(length(variant.id), 5))

# get genotypic data
seqGetData(f, "genotype")



## set the intersection

seqResetFilter(f)
seqSetFilterChrom(f, 10L)
seqSummary(f, "genotype", check="none")

AF &lt;- seqAlleleFreq(f)
table(AF &lt;= 0.9)

seqSetFilter(f, variant.sel=(AF&lt;=0.9), action="intersect")
seqSummary(f, "genotype", check="none")



## chromosome

seqResetFilter(f)

seqSetFilterChrom(f, is.num=TRUE)
seqSummary(f, "genotype", check="none")

seqSetFilterChrom(f, is.num=FALSE)
seqSummary(f, "genotype", check="none")

seqSetFilterChrom(f, 1:4)
seqSummary(f, "genotype", check="none")
table(seqGetData(f, "chromosome"))

# HLA region
seqSetFilterChrom(f, 6, from.bp=29719561, to.bp=32883508)
seqSummary(f, "genotype", check="none")

# two regions
seqSetFilterChrom(f, c(1, 6), from.bp=c(1000000, 29719561),
    to.bp=c(90000000, 32883508))
seqSummary(f, "genotype", check="none")
seqGetData(f, "chromosome")



# close the GDS file
seqClose(f)
</pre>

<hr /><div style="text-align: center;">[Package <em>SeqArray</em> version 1.13.6 <a href="00Index.html">Index</a>]</div>
</body></html>
