<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Parse the Header of a VCF File</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for seqVCF_Header {SeqArray}"><tr><td>seqVCF_Header {SeqArray}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Parse the Header of a VCF File</h2>

<h3>Description</h3>

<p>Parses the header of a VCF file.
</p>


<h3>Usage</h3>

<pre>
seqVCF_Header(vcf.fn, getnum=FALSE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>vcf.fn</code></td>
<td>
<p>the file name; or a <code><a href="../../base/html/connections.html">connection</a></code> object</p>
</td></tr>
<tr valign="top"><td><code>getnum</code></td>
<td>
<p>if <code>TRUE</code>, return the number of samples and variants</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The ID description contains four columns: <code>ID</code> &ndash; variable name;
<code>Number</code> &ndash; the number of elements, see the webpage of the 1000 Genomes
Project; <code>Type</code> &ndash; data type; <code>Description</code> &ndash; a variable
description.
</p>


<h3>Value</h3>

<p>Return a list (with a class name &quot;SeqVCFHeaderClass&quot;, S3 object):
</p>
<table summary="R valueblock">
<tr valign="top"><td><code>fileformat</code></td>
<td>
<p>the file format</p>
</td></tr>
<tr valign="top"><td><code>info</code></td>
<td>
<p>the ID description in the INFO field</p>
</td></tr>
<tr valign="top"><td><code>filter</code></td>
<td>
<p>the ID description in the FILTER field</p>
</td></tr>
<tr valign="top"><td><code>format</code></td>
<td>
<p>the ID description in the FORMAT field</p>
</td></tr>
<tr valign="top"><td><code>alt</code></td>
<td>
<p>the ID description in the ALT field</p>
</td></tr>
<tr valign="top"><td><code>contig</code></td>
<td>
<p>the description in the contig field</p>
</td></tr>
<tr valign="top"><td><code>assembly</code></td>
<td>
<p>the link of assembly</p>
</td></tr>
<tr valign="top"><td><code>reference</code></td>
<td>
<p>genome reference, or <code>NULL</code> if unknown</p>
</td></tr>
<tr valign="top"><td><code>header</code></td>
<td>
<p>the other header lines</p>
</td></tr>
<tr valign="top"><td><code>ploidy</code></td>
<td>
<p>ploidy, two for humans</p>
</td></tr>
<tr valign="top"><td><code>num.sample</code></td>
<td>
<p>the number of samples</p>
</td></tr>
<tr valign="top"><td><code>num.variant</code></td>
<td>
<p>the number of variants</p>
</td></tr>
</table>


<h3>Author(s)</h3>

<p>Xiuwen Zheng</p>


<h3>References</h3>

<p>Danecek, P., Auton, A., Abecasis, G., Albers, C.A., Banks, E., DePristo,
M.A., Handsaker, R.E., Lunter, G., Marth, G.T., Sherry, S.T., et al. (2011).
The variant call format and VCFtools. Bioinformatics 27, 2156-2158.
</p>


<h3>See Also</h3>

<p><code><a href="seqVCF_SampID.html">seqVCF_SampID</a></code>, <code><a href="seqVCF2GDS.html">seqVCF2GDS</a></code>
</p>


<h3>Examples</h3>

<pre>
# the VCF file
(vcf.fn &lt;- seqExampleFileName("vcf"))
# or vcf.fn &lt;- "C:/YourFolder/Your_VCF_File.vcf"

# get sample id
seqVCF_Header(vcf.fn, getnum=TRUE)

# use a connection object
f &lt;- file(vcf.fn, "r")
seqVCF_Header(f, getnum=TRUE)
close(f)
</pre>

<hr /><div style="text-align: center;">[Package <em>SeqArray</em> version 1.17.1 <a href="00Index.html">Index</a>]</div>
</body></html>
